README.md
# DebounceAndThrottle

[![Hex version badge](https://img.shields.io/hexpm/v/debounce_and_throttle.svg?color=3b4)](https://hex.pm/packages/debounce_and_throttle)
[![License](https://img.shields.io/badge/license-MIT-green?color=3b4)](https://github.com/johnknott/debounce-and-throttle/blob/master/LICENSE.md)
[![Code coverage badge](https://img.shields.io/codecov/c/github/johnknott/debounce-and-throttle/badge.svg?color=3b4)](https://app.codecov.io/gh/johnknott/debounce-and-throttle)
[![Github Workflow](https://img.shields.io/github/workflow/status/johnknott/debounce-and-throttle/Elixir%20CI?logo=github&color=3b4)](https://github.com/johnknott/debounce-and-throttle/actions/workflows/elixir.yml)

DebounceAndThrottle is a simple library to allow to *debounce* or *throttle* function calls or message sending.

The following page explains throttling and debouncing quite well, albeit from a javascript perspective.

[https://css-tricks.com/debouncing-throttling-explained-examples/]

Examples can be found below and full documentation can be found at [hexdocs.pm](https://hexdocs.pm/debounce_and_throttle/api-reference.html).

## Installation

The package can be installed by adding `debounce_and_throttle` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:debounce_and_throttle, "~> 0.9.0"}
  ]
end
```

## Setup

### Setup with a Supervision Tree, perhaps in a Phoenix project

Add `DebounceAndThrottle.Server` to your Supervision tree in `lib/<application_name>/application.ex`

```elixir
children = [DebounceAndThrottle.Server, ...]
```

### Setup without a Supervision Tree, perhaps in a standalone script

```elixir
DebounceAndThrottle.Server.start_link([])
```

## Usage

Alias Debounce and Throttle if you want to keep things terse.

```elixir
alias DebounceAndThrottle.{Debounce, Throttle}
```

### Debounce

This anonymous function will be called after 5 seconds, but only if this method wasnt called in the interim with the same `key`

```elixir
Debounce.call(fn -> IO.puts("Hey there!") end, "some-key", 5_000)
```

This module, function and arguments will be called after 5 seconds, but only if this method wasnt called in the interim with the same `key`

```elixir
Debounce.apply(IO, :puts, ["Hey there!"], "some-key", 5_000)
```

This message will be sent after 5 seconds, but only if this method wasnt called in the interim with the same `key`

```elixir
Debounce.send(self(), :some_message, "some-key", 5_000)
```

Returns the state - the current list of debounced functions. Useful for debugging.

```elixir
iex> Debounce.state
%{
  apply: %{},
  call: %{
    "say_hey" => %DebounceAndThrottle.Debounce{
      debounced_count: 1,
      extra_data: %{fun: 'Function<45.65746770/0 in :erl_eval.expr/5>'},
      scheduled_at: ~U[2022-03-12 22:50:01.190171Z],
      timer_ref: 'Reference<0.418177534.3850108929.259344>'
    }
  },
  send: %{}
}
```

### Throttle

This anonymous function will be called straight away, but only if this method wasnt called in the last 5 seconds with the same `key`

```elixir
Throttle.call(fn -> IO.puts("Hey there!") end, "some-key", 5_000)
```

This module, function and arguments will be called straight away, but only if this method wasnt called in last 5 seconds with the same `key`

```elixir
Throttle.apply(IO, :puts, ["Hey there!"], "some-key", 5_000)
```

This message will be sent straight away, but only if this method wasnt called in the last 5 seconds with the same `key`

```elixir
Throttle.send(self(), :some_message, "some-key", 5_000)
```

Returns the state - the current list of throttled functions. Useful for debugging.

```elixir
iex> Throttle.state
%{
  apply: %{},
  call: %{
    "say_hey" => %DebounceAndThrottle.Throttle{
      extra_data: %{fun: 'Function<45.65746770/0 in :erl_eval.expr/5>'},
      status: :executed,
      throttled_count: 0,
      throttled_until: -576460730743,
      throttled_until_utc: ~U[2022-03-12 22:47:39.829107Z]
    }
    ...
  },
  send: %{}
}
```

## Notes

### Choice of key

Your choice of what to use as the `key` depends on the granularity you want to throttle or debounce things.

So, bad example - if you wanted to throttle a `send_sms` function - if you used `send_sms` as the key it would throttle the function for every user in the system which is probably not what you want. So you might use something like `sms-#{user_id}` as the key to throttle it per user.

## License

This project is licensed under the MIT License - see the LICENSE.md file for details


CHANGELOG.md
# Changelog

I'll try to follow the format of: [https://keepachangelog.com/en/1.0.0/]

## Unreleased

## [0.9.0] - 2022-04-12

### Added

- Initial release.
- Core functionality all integrated.
- Added full test coverage.
- Added credo.
- Added coveralls.
- Added Github actions and shields.io badges.


test/server_test.exs
defmodule ServerTest do
  use ExUnit.Case
  alias DebounceAndThrottle.{Throttle, Server}

  setup do
    Server.reset()
    :ok
  end

  test "Server.reset works properly" do
    assert(:sys.get_state(Server) == Server.initial_state())
    Throttle.call(fn -> send(self(), :some_message) end, "Throttle.call.1", 10_000)
    assert(:sys.get_state(Server) != Server.initial_state())
    Server.reset()
    assert(:sys.get_state(Server) == Server.initial_state())
  end

  test "Server.cleanup works properly" do
    assert(:sys.get_state(Server) == Server.initial_state())
    Throttle.call(fn -> send(self(), :some_message) end, "Throttle.call.1", -10_000)
    assert(:sys.get_state(Server) != Server.initial_state())
    new_state = Server.cleanup(:sys.get_state(Server), [:call, :apply, :send])
    assert(new_state == Server.initial_state())
  end

  test "Test unknown message doesnt crash Genserver" do
    send(Server, :unknown_message)
    :timer.sleep(100)
    assert(is_map(DebounceAndThrottle.Debounce.state()))
  end
end


test/debounce_test.exs
defmodule DebounceTest do
  use ExUnit.Case
  alias DebounceAndThrottle.{Debounce, Server}

  doctest Debounce

  setup do
    Server.reset()
    :ok
  end

  test "Debounce.send schedules an event correctly" do
    delay = 5_000
    should_be_scheduled_at = DateTime.add(DateTime.utc_now(), delay, :millisecond)

    assert {:ok,
            %{
              debounced_count: 1,
              extra_data: %{message: :some_message, pid: Server},
              scheduled_at: time,
              timer_ref: timer_ref
            }} = Debounce.send(Server, :some_message, "Debounce.send.1", delay)

    assert(is_reference(timer_ref))
    assert(is_struct(time))
    assert(Time.diff(time, should_be_scheduled_at, :millisecond) < 100)

    assert {:ok,
            %{
              debounced_count: 2,
              extra_data: %{message: :some_message, pid: Server},
              scheduled_at: new_time,
              timer_ref: new_timer_ref
            }} = Debounce.send(Server, :some_message, "Debounce.send.1", delay)

    assert(timer_ref != new_timer_ref)
    assert(time != new_time)
    refute_receive(:some_message, 100)

    assert {:ok, %{debounced_count: 1}} = Debounce.send(Server, :some_message, "different_key", delay)
  end

  test "Debounce.send executes the command correctly" do
    delay = 100
    Debounce.send(self(), :some_message, "Debounce.send.1", delay)
    assert_receive :some_message, 1_000
  end

  test "Debounce.call schedules an event correctly" do
    delay = 5_000
    should_be_scheduled_at = DateTime.add(DateTime.utc_now(), delay, :millisecond)

    assert {:ok,
            %{
              debounced_count: 1,
              extra_data: %{fun: fun},
              scheduled_at: time,
              timer_ref: timer_ref
            }} = Debounce.call(fn -> IO.puts("Hey1") end, "Debounce.call.1", delay)

    assert(is_reference(timer_ref))
    assert(is_struct(time))
    assert(Time.diff(time, should_be_scheduled_at, :millisecond) < 100)
    assert(is_function(fun))

    assert {:ok,
            %{
              debounced_count: 2,
              scheduled_at: new_time,
              timer_ref: new_timer_ref
            }} = Debounce.call(fn -> IO.puts("Hey1") end, "Debounce.call.1", delay)

    assert(timer_ref != new_timer_ref)
    assert(time != new_time)
    refute_receive(:some_message, 100)

    assert {:ok, %{debounced_count: 1}} = Debounce.call(fn -> IO.puts("Hey2") end, "Debounce.call.2", delay)
  end

  test "Debounce.call executes the function correctly" do
    delay = 100
    this = self()

    Debounce.call(fn -> send(this, :some_message) end, "Debounce.call.3", delay)

    assert_receive :some_message, 1_000
  end

  test "Debounce.apply schedules a function correctly" do
    delay = 5_000
    should_be_scheduled_at = DateTime.add(DateTime.utc_now(), delay, :millisecond)

    this = self()

    assert {:ok,
            %{
              debounced_count: 1,
              extra_data: %{module: Process, fun: :send, args: [^this, :some_message, []]},
              scheduled_at: time,
              timer_ref: timer_ref
            }} = Debounce.apply(Process, :send, [self(), :some_message, []], "Debounce.apply.1", delay)

    assert(is_reference(timer_ref))
    assert(is_struct(time))
    assert(Time.diff(time, should_be_scheduled_at, :millisecond) < 100)

    assert {:ok,
            %{
              debounced_count: 2,
              scheduled_at: new_time,
              timer_ref: new_timer_ref
            }} = Debounce.apply(Process, :send, [self(), :some_message, []], "Debounce.apply.1", delay)

    assert(timer_ref != new_timer_ref)
    assert(time != new_time)
    refute_receive(:some_message, 100)

    assert {:ok, %{debounced_count: 1}} = Debounce.call(fn -> IO.puts("Hey2") end, "Debounce.apply.2", delay)
  end

  test "Debounce.apply executes a function correctly" do
    delay = 100

    Debounce.apply(Process, :send, [self(), :some_message, []], "Debounce.apply.1", delay)
    assert_receive :some_message, 1_000
  end

  test "Debounce.state returns correct state" do
    state = Debounce.state()
    assert(state == Server.initial_state()[:debounced])
    Debounce.apply(Process, :send, [self(), :some_message, []], "Debounce.apply.1", 100)
    state = Debounce.state()
    assert(length(Map.keys(state[:apply])) == 1)
    assert(Map.keys(state[:send]) == [])
    assert(Map.keys(state[:call]) == [])
    key_state = get_in(state, [:apply, "Debounce.apply.1"])
    assert(%Debounce{} = key_state)
  end
end


test/throttle_test.exs
defmodule ThrottleTest do
  use ExUnit.Case
  alias DebounceAndThrottle.{Throttle, Server}

  doctest Throttle

  setup do
    Server.reset()
    :ok
  end

  test "Throttle.send sends and then throttles a message correctly" do
    delay = 5_000
    should_be_throttled_until = DateTime.add(DateTime.utc_now(), delay, :millisecond)

    this = self()

    assert {:ok,
            %{
              throttled_count: 0,
              status: :executed,
              extra_data: %{message: :some_message, pid: ^this},
              throttled_until: until,
              throttled_until_utc: until_utc
            }} = Throttle.send(self(), :some_message, "Throttle.send.1", delay)

    assert(is_number(until))
    assert(is_struct(until_utc))
    assert(Time.diff(until_utc, should_be_throttled_until, :millisecond) < 100)
    assert_receive(:some_message, 100)

    assert {:ok,
            %{
              throttled_count: 1,
              status: :throttled,
              extra_data: %{message: :some_message, pid: ^this},
              throttled_until: new_until,
              throttled_until_utc: new_until_utc
            }} = Throttle.send(self(), :some_message, "Throttle.send.1", delay)

    assert(until == new_until)
    assert(until_utc == new_until_utc)
    refute_receive(:some_message, 100)

    assert {:ok, %{throttled_count: 0, status: :executed}} =
             Throttle.send(self(), :some_message, "different_key", delay)

    assert_receive :some_message, 100
  end

  test "Throttle.call calls and then throttles a function correctly" do
    delay = 5_000
    should_be_throttled_until = DateTime.add(DateTime.utc_now(), delay, :millisecond)

    this = self()

    assert {:ok,
            %{
              throttled_count: 0,
              status: :executed,
              extra_data: %{fun: fun},
              throttled_until: until,
              throttled_until_utc: until_utc
            }} = Throttle.call(fn -> send(this, :some_message) end, "Throttle.call.1", delay)

    assert(is_function(fun))
    assert(is_number(until))
    assert(is_struct(until_utc))
    assert(Time.diff(until_utc, should_be_throttled_until, :millisecond) < 100)
    assert_receive(:some_message, 100)

    assert {:ok,
            %{
              throttled_count: 1,
              status: :throttled,
              extra_data: %{fun: new_fun},
              throttled_until: new_until,
              throttled_until_utc: new_until_utc
            }} = Throttle.call(fn -> send(this, :some_message) end, "Throttle.call.1", delay)

    refute_receive(:some_message, 100)

    assert(fun == new_fun)
    assert(until == new_until)
    assert(until_utc == new_until_utc)

    assert {:ok, %{throttled_count: 0, status: :executed}} =
             Throttle.call(fn -> send(this, :some_message) end, "different_key", delay)

    assert_receive(:some_message, 100)
  end

  test "Throttle.apply schedules a function correctly" do
    delay = 5_000
    should_be_throttled_until = DateTime.add(DateTime.utc_now(), delay, :millisecond)

    this = self()

    assert {:ok,
            %{
              throttled_count: 0,
              status: :executed,
              extra_data: %{module: Process, fun: :send, args: [^this, :some_message, []]},
              throttled_until: until,
              throttled_until_utc: until_utc
            }} = Throttle.apply(Process, :send, [self(), :some_message, []], "Throttle.apply.1", delay)

    assert(is_number(until))
    assert(is_struct(until_utc))
    assert(Time.diff(until_utc, should_be_throttled_until, :millisecond) < 100)
    assert_receive(:some_message, 100)

    assert {:ok,
            %{
              throttled_count: 1,
              status: :throttled,
              extra_data: %{module: Process, fun: :send, args: [^this, :some_message, []]},
              throttled_until: new_until,
              throttled_until_utc: new_until_utc
            }} = Throttle.apply(Process, :send, [self(), :some_message, []], "Throttle.apply.1", delay)

    assert(until == new_until)
    assert(until_utc == new_until_utc)

    assert {:ok, %{throttled_count: 0, status: :executed}} =
             Throttle.call(fn -> send(this, :some_message) end, "different_key", delay)

    assert_receive(:some_message, 100)
  end

  test "Throttle.apply executes a function correctly" do
    delay = 100
    Throttle.apply(Process, :send, [self(), :some_message, []], "Debounce.apply.1", delay)
    assert_receive :some_message, 1_000
  end

  test "Throttle.state returns correct state" do
    state = Throttle.state()
    assert(state == Server.initial_state()[:throttled])
    Throttle.apply(Process, :send, [self(), :some_message, []], "Throttle.apply.1", 100)
    state = Throttle.state()
    assert(length(Map.keys(state[:apply])) == 1)
    assert(Map.keys(state[:send]) == [])
    assert(Map.keys(state[:call]) == [])
    key_state = get_in(state, [:apply, "Throttle.apply.1"])
    assert(%Throttle{} = key_state)
  end
end


mix.exs
defmodule DebounceAndThrottle.MixProject do
  use Mix.Project
  @github_url "https://github.com/johnknott/debounce-and-throttle"

  def project do
    [
      app: :debounce_and_throttle,
      version: "0.9.0",
      name: "Debounce and Throttle",
      description: "Provides simple debounce and throttle functionality",
      source_url: @github_url,
      homepage_url: @github_url,
      files: ~w(mix.exs lib LICENSE.md README.md CHANGELOG.md),
      package: [
        maintainers: ["John Knott"],
        licenses: ["MIT"],
        links: %{
          "GitHub" => @github_url
        }
      ],
      docs: [
        main: "readme",
        extras: ["README.md", "CHANGELOG.md"]
      ],
      elixir: "~> 1.13",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test
      ]
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger],
      mod: {DebounceAndThrottle.Application, []}
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:credo, "~> 1.6", only: [:dev, :test], runtime: false},
      {:excoveralls, "~> 0.10", only: :test},
      {:ex_doc, "~> 0.27", only: :dev, runtime: false},
      {:dialyxir, "~> 1.0", only: [:dev], runtime: false}
    ]
  end
end


lib/debounce_and_throttle/server.ex
defmodule DebounceAndThrottle.Server do
  @moduledoc """
  The Genserver that implements the Debounce and Throttle functionality.
  End users will not use this, instead using the API in debounce.ex and throttle.ex
  """

  use GenServer
  alias DebounceAndThrottle.{Throttle, Debounce}
  require Logger

  # Client

  def start_link(state) do
    GenServer.start_link(__MODULE__, state, name: __MODULE__)
  end

  def reset() do
    GenServer.call(__MODULE__, :reset)
  end

  # Server (callbacks)

  @impl true
  def init(_state) do
    :timer.send_interval(1_000, :cleanup)
    {:ok, initial_state()}
  end

  @impl true
  def handle_call(:reset, _, state) do
    cancel_timers(state)
    {:reply, :ok, initial_state()}
  end

  @impl true
  def handle_call({:state, type}, _, state) do
    {:reply, state[type], state}
  end

  @impl true
  def handle_call({:send_debounced, {pid, message, key, period}}, _, state) do
    old_key_state = state[:debounced][:send][key]
    {timer_ref, scheduled_at, count} = set_timer(old_key_state, {:send_fun, pid, message, key}, period)
    key_state = debounced_key_state(timer_ref, scheduled_at, count, %{pid: pid, message: message})
    {:reply, key_state, add_fun_state(state, :debounced, :send, key, key_state)}
  end

  @impl true
  def handle_call({:call_debounced, {fun, key, period}}, _, state) do
    old_key_state = state[:debounced][:call][key]
    {timer_ref, scheduled_at, count} = set_timer(old_key_state, {:call_fun, fun, key}, period)
    key_state = debounced_key_state(timer_ref, scheduled_at, count, %{fun: fun})
    {:reply, key_state, add_fun_state(state, :debounced, :call, key, key_state)}
  end

  @impl true
  def handle_call({:apply_debounced, {module, fun, args, key, period}}, _, state) do
    old_key_state = state[:debounced][:apply][key]
    {timer_ref, scheduled_at, count} = set_timer(old_key_state, {:apply_fun, module, fun, args, key}, period)
    key_state = debounced_key_state(timer_ref, scheduled_at, count, %{module: module, fun: fun, args: args})
    {:reply, key_state, add_fun_state(state, :debounced, :apply, key, key_state)}
  end

  @impl true
  def handle_call({:send_throttled, {pid, message, key, period}}, _, state) do
    old_key_state = state[:throttled][:send][key]
    key_state = throttled_key_state(old_key_state, period, %{pid: pid, message: message})
    if key_state.status == :executed, do: send(pid, message)
    {:reply, key_state, add_fun_state(state, :throttled, :send, key, key_state)}
  end

  @impl true
  def handle_call({:call_throttled, {fun, key, period}}, _, state) do
    old_key_state = state[:throttled][:call][key]
    key_state = throttled_key_state(old_key_state, period, %{fun: fun})
    if key_state.status == :executed, do: spawn(fun)
    {:reply, key_state, add_fun_state(state, :throttled, :call, key, key_state)}
  end

  @impl true
  def handle_call({:apply_throttled, {module, fun, args, key, period}}, _, state) do
    old_key_state = state[:throttled][:apply][key]
    key_state = throttled_key_state(old_key_state, period, %{module: module, fun: fun, args: args})
    if key_state.status == :executed, do: spawn(fn -> apply(module, fun, args) end)
    {:reply, key_state, add_fun_state(state, :throttled, :apply, key, key_state)}
  end

  @impl true
  def handle_info({:send_fun, pid, message, key}, state) do
    send(pid, message)
    {:noreply, remove_fun_state(state, :send, key)}
  end

  @impl true
  def handle_info({:call_fun, fun, key}, state) do
    spawn(fun)
    {:noreply, remove_fun_state(state, :call, key)}
  end

  @impl true
  def handle_info({:apply_fun, module, fun, args, key}, state) do
    spawn(fn -> apply(module, fun, args) end)
    {:noreply, remove_fun_state(state, :apply, key)}
  end

  @impl true
  def handle_info(:cleanup, state) do
    {:noreply, cleanup(state, [:call, :apply, :send])}
  end

  @impl true
  def handle_info(message, state) do
    Logger.debug("#{__MODULE__}: Unknown message: #{inspect(message)}")
    {:noreply, state}
  end

  # Helpers

  def initial_state() do
    sca = %{send: %{}, call: %{}, apply: %{}}
    %{debounced: sca, throttled: sca}
  end

  def cleanup(state, types) do
    Enum.reduce(types, state, fn type, new_state ->
      filtered =
        :maps.filter(
          fn _, v -> v.throttled_until > System.monotonic_time(:millisecond) end,
          state[:throttled][type]
        )

      put_in(new_state, [:throttled, type], filtered)
    end)
  end

  defp cancel_timers(state) do
    for type <- [:apply, :call, :send] do
      get_in(state, [:debounced, type]) |> Map.values() |> Enum.each(fn x -> Process.cancel_timer(x.timer_ref) end)
    end
  end

  defp run_at(period), do: DateTime.add(DateTime.utc_now(), period, :millisecond)

  defp set_timer(%{timer_ref: timer_ref, debounced_count: debounced_count}, message, period) do
    Process.cancel_timer(timer_ref)
    {Process.send_after(self(), message, period), run_at(period), debounced_count + 1}
  end

  defp set_timer(nil, message, period), do: {Process.send_after(self(), message, period), run_at(period), 1}

  defp add_fun_state(state, ns = :debounced, type, key, key_state),
    do: put_in(state, [ns, type, key], key_state)

  defp add_fun_state(state, ns = :throttled, type, key, key_state),
    do: put_in(state, [ns, type, key], key_state)

  defp remove_fun_state(state, type, key), do: pop_in(state, [:debounced, type, key]) |> elem(1)

  defp debounced_key_state(timer_ref, scheduled_at, debounced_count, extra_data) do
    %Debounce{
      timer_ref: timer_ref,
      scheduled_at: scheduled_at,
      debounced_count: debounced_count,
      extra_data: extra_data
    }
  end

  defp throttled_key_state(old_key_state, period, extra_data) do
    now = System.monotonic_time(:millisecond)

    if old_key_state == nil or now > old_key_state.throttled_until do
      # Not throttled, or was throttled but now expired
      %Throttle{
        status: :executed,
        throttled_until: now + period,
        throttled_until_utc: DateTime.add(DateTime.utc_now(), period, :millisecond),
        throttled_count: 0,
        extra_data: extra_data
      }
    else
      # Throttled
      Map.update(%{old_key_state | status: :throttled}, :throttled_count, 0, fn old_count ->
        old_count + 1
      end)
    end
  end
end


lib/debounce_and_throttle/debounce.ex
defmodule DebounceAndThrottle.Debounce do
  defstruct([:timer_ref, :scheduled_at, :debounced_count, :extra_data])
  alias DebounceAndThrottle.Debounce

  @type t :: %Debounce{
          timer_ref: reference(),
          scheduled_at: DateTime.t(),
          debounced_count: non_neg_integer(),
          extra_data: map()
        }

  @moduledoc """
  This module implements the Debounce API.
  """
  @server DebounceAndThrottle.Server

  @doc """
  Sends a `message` to a given `pid`, but only after `period` has passed without any more calls to this function with the same `key`

  Returns `{:ok, %Debounce{}}`.
  """
  @spec send(pid() | atom(), term(), String.t(), non_neg_integer()) :: {:ok, Debounce.t()}
  def send(pid, message, key, period) do
    result = GenServer.call(@server, {:send_debounced, {pid, message, key, period}})
    {:ok, result}
  end

  @doc """
  Calls a `fun` but only after `period` has passed without any more calls to this function with the same `key`

  Returns `{:ok, %Debounce{}}`.
  """
  @spec call(fun(), String.t(), non_neg_integer()) :: {:ok, Debounce.t()}
  def call(fun, key, period) when is_function(fun) do
    result = GenServer.call(@server, {:call_debounced, {fun, key, period}})
    {:ok, result}
  end

  @doc """
  Calls a `fun` but only after `period` has passed without any more calls to this function with the same `key`

  Returns `{:ok, %Debounce{}}`.
  """
  @spec apply(module, fun :: atom(), [any], String.t(), non_neg_integer()) :: {:ok, Debounce.t()}
  def apply(module, fun, args, key, period) do
    result = GenServer.call(@server, {:apply_debounced, {module, fun, args, key, period}})
    {:ok, result}
  end

  @doc """
  Returns the state - the current list of debounced functions. Useful for debugging.

  Returns something like:
  %{
    apply: %{},
    call: %{
      "say_hey" => %DebounceAndThrottle.Debounce{
        debounced_count: 1,
        extra_data: %{fun: #Function<45.65746770/0 in :erl_eval.expr/5>},
        scheduled_at: ~U[2022-03-12 22:50:01.190171Z],
        timer_ref: #Reference<0.418177534.3850108929.259344>
      }
    },
    send: %{}
  }
  """
  @spec state() :: map()
  def state(), do: GenServer.call(@server, {:state, :debounced})
end


lib/debounce_and_throttle/throttle.ex
defmodule DebounceAndThrottle.Throttle do
  alias DebounceAndThrottle.Throttle
  defstruct([:status, :throttled_until, :throttled_until_utc, :throttled_count, :extra_data])

  @type t :: %Throttle{
          status: atom(),
          throttled_until: integer(),
          throttled_until_utc: DateTime.t(),
          throttled_count: non_neg_integer(),
          extra_data: map()
        }

  @moduledoc """
  This module implements the Throttle API.
  """
  @server DebounceAndThrottle.Server

  @doc """
  Sends a `message` to a given `pid`, but only once during `period` per `key`

  Returns `{:ok, %Throttle{}}`.
  """
  @spec send(pid() | atom(), term(), String.t(), non_neg_integer()) :: {:ok, Throttle.t()}
  def send(pid, message, key, period) do
    result = GenServer.call(@server, {:send_throttled, {pid, message, key, period}})
    {:ok, result}
  end

  @doc """
  Calls a `fun`, but only once during `period` per `key`

  Returns `{:ok, %Throttle{}}`.
  """
  @spec call(fun(), String.t(), non_neg_integer()) :: {:ok, Throttle.t()}
  def call(fun, key, period) do
    result = GenServer.call(@server, {:call_throttled, {fun, key, period}})
    {:ok, result}
  end

  @doc """
  Calls a `func`, but only once during `period` per `key`

  Returns `{:ok, %Throttle{}}`.
  """
  @spec apply(module, fun :: atom(), [any], String.t(), non_neg_integer()) :: {:ok, Throttle.t()}
  def apply(module, fun, args, key, period) do
    result = GenServer.call(@server, {:apply_throttled, {module, fun, args, key, period}})
    {:ok, result}
  end

  @doc """
  Returns the state - the current list of throttled functions. Useful for debugging.

  Returns something like:
  %{
    apply: %{},
    call: %{
      "say_hey" => %DebounceAndThrottle.Throttle{
        extra_data: %{fun: #Function<45.65746770/0 in :erl_eval.expr/5>},
        status: :executed,
        throttled_count: 0,
        throttled_until: -576460730743,
        throttled_until_utc: ~U[2022-03-12 22:47:39.829107Z]
      }
      ...
    },
    send: %{}
  }
  """
  @spec state() :: map()
  def state(), do: GenServer.call(@server, {:state, :throttled})
end
